miner/worker.go

```diff
@@ -1062,39 +1062,35 @@ loop:
-			log.Info("Skipping L1 message", "queueIndex", queueIndex, "tx", tx.Hash().String(), "block", w.current.header.Number, "reason", "row consumption overflow")
-			w.current.nextL1MsgIndex = queueIndex + 1
-
-			// after `ErrTxRowConsumptionOverflow`, ccc might not revert updates
-			// associated with this transaction so we cannot pack more transactions.
-			// TODO: fix this in ccc and change these lines back to `txs.Shift()`
-			circuitCapacityReached = true
-			break loop
-
-		case (errors.Is(err, circuitcapacitychecker.ErrTxRowConsumptionOverflow) && !tx.IsL1MessageTx()):
-			// Circuit capacity check: L2MessageTx row consumption too high, skip the account.
-			// This is also useful for skipping "problematic" L2MessageTxs.
-			log.Trace("Circuit capacity limit reached for a single tx", "tx", tx.Hash().String())
-
-			// after `ErrTxRowConsumptionOverflow`, ccc might not revert updates
-			// associated with this transaction so we cannot pack more transactions.
-			// TODO: fix this in ccc and change these lines back to `txs.Pop()`
-			circuitCapacityReached = true
-			break loop
+			// Circuit capacity check.
+			if w.current.tcount >= 1 { // 1. circuit capacity limit reached in a block, and it's not the first tx:
+				// don't pop or shift, just quit the loop immediately;
+				// though it might still be possible to add some "smaller" txs,
+				// but it's a trade-off between tracing overhead & block usage rate
+				log.Trace("Circuit capacity limit reached in a block", "acc_rows", w.current.accRows, "tx", tx.Hash().String())
+				circuitCapacityReached = true
+				if tx.IsL1MessageTx() {
+					queueIndex := tx.AsL1MessageTx().QueueIndex
+					log.Info("Skipping L1 message", "queueIndex", queueIndex, "tx", tx.Hash().String(), "block", w.current.header.Number, "reason", "row consumption overflow")
+					w.current.nextL1MsgIndex = queueIndex + 1
+				}
+				break loop
+			} else { // 2. circuit capacity limit reached in a block, and it's the first tx: drop the tx.
+				circuitCapacityReached = false
+				if tx.IsL1MessageTx() { // 2.1 L1MessageTx row consumption too high, shift to the next from the account,
+					// because we shouldn't skip the entire txs from the same account.
+					// This is also useful for skipping "problematic" L1MessageTxs.
+					queueIndex := tx.AsL1MessageTx().QueueIndex
+					log.Trace("Circuit capacity limit reached for a single tx", "tx", tx.Hash().String(), "queueIndex", queueIndex)
+					log.Info("Skipping L1 message", "queueIndex", queueIndex, "tx", tx.Hash().String(), "block", w.current.header.Number, "reason", "row consumption overflow")
+					w.current.nextL1MsgIndex = queueIndex + 1
+					txs.Shift()
+				} else { // 2.2 L2MessageTx row consumption too high, skip the account.
+					// This is also useful for skipping "problematic" L2MessageTxs.
+					log.Trace("Circuit capacity limit reached for a single tx", "tx", tx.Hash().String())
+					txs.Pop()
+				}
+			}
```

Comment:
Do we need to call `ccc.Reset()` here?

Reply:
Hi @HAOYUatHZ , I check [the code of capacity-checker](https://github.com/scroll-tech/scroll-prover/blob/main/prover/src/zkevm/capacity_checker.rs#L189) in scroll-prover. Seems we need to reset it when skips the first tx, otherwise it would always be `ErrBlockRowConsumptionOverflow`. Please correct me if I was wrong.

I add ccc reset in commit [325e045](https://github.com/scroll-tech/go-ethereum/pull/475/commits/325e0452dcec4fa293d73b4dfc4e6a6b1f722f91).